File: src/main/java/com/github/tammo/fun/frontend/type/TypeAnnotate.java
package com.github.tammo.fun.frontend.type;

import com.github.tammo.fun.frontend.ast.Expression;
import com.github.tammo.fun.frontend.ast.SyntaxNode;

import java.util.List;

public class TypeAnnotate {

    public static TypedTreeNode annotateTypes(SyntaxNode syntaxNode) {
        return switch (syntaxNode) {
            case SyntaxNode.ClassDeclaration classDeclaration -> {
                List<TypedTreeNode.EffectDeclaration> effectDeclarations = classDeclaration.effectDeclarations()
                        .stream()
                        .map(TypeAnnotate::annotateTypes)
                        .map(TypedTreeNode.EffectDeclaration.class::cast)
                        .toList();
                yield new TypedTreeNode.ObjectDeclaration(
                        classDeclaration.name(),
                        effectDeclarations
                );
            }
            case SyntaxNode.CompilationUnit compilationUnit -> new TypedTreeNode.CompilationUnit(
                    ((TypedTreeNode.ObjectDeclaration) annotateTypes(compilationUnit.classDeclaration()))
            );
            case SyntaxNode.EffectDeclaration effectDeclaration -> {
                List<TypedTreeNode.Parameter> typedParameter = effectDeclaration.parameters().parameters().stream()
                        .map(TypeAnnotate::annotateTypes)
                        .map(TypedTreeNode.Parameter.class::cast)
                        .toList();
                TypedTreeNode.Expression typedBody = (TypedTreeNode.Expression) annotateTypes(effectDeclaration.body());
                yield new TypedTreeNode.EffectDeclaration(
                        new TypedTreeNode.TypedIdentifier(effectDeclaration.name(), new FunType.TypeVariable()),
                        typedParameter,
                        typedBody
                );
            }
            case SyntaxNode.ExposeDeclaration ignore -> new TypedTreeNode.NotImplemented();
            case SyntaxNode.NamespaceDeclaration namespaceDeclaration ->
                    new TypedTreeNode.NamespaceDeclaration(namespaceDeclaration.qualifierIdentifier());
            case SyntaxNode.UseDeclaration ignore -> new TypedTreeNode.NotImplemented();
            case SyntaxNode.Parameter parameter -> new TypedTreeNode.Parameter(
                    new TypedTreeNode.TypedIdentifier(parameter.name(), FunType.fromString(parameter.type()))
            );
            case SyntaxNode.ParameterList ignored -> new TypedTreeNode.NotImplemented();
            case SyntaxNode.FunctionDeclaration ignored -> new TypedTreeNode.NotImplemented();
            case Expression expression -> annotateTypesAtExpression(expression);
        };
    }

    private static TypedTreeNode.Expression annotateTypesAtExpression(Expression expression) {
        return switch (expression) {
            case Expression.ArithmeticExpression arithmeticExpression ->
                    anonotateTypesAtArithmeticExpression(arithmeticExpression);
            case Expression.Term term -> annotateTypesAtTerm(term);
            case Expression.Factor factor -> annotateTypesAtFactor(factor);
            case Expression.FunctionCall functionCall -> {
                TypedTreeNode.TypedIdentifier typedIdentifier =
                        new TypedTreeNode.TypedIdentifier(functionCall.functionName(), new FunType.TypeVariable());
                List<TypedTreeNode.Expression> arguments = functionCall.arguments().stream()
                        .map(TypeAnnotate::annotateTypesAtExpression)
                        .toList();
                yield new TypedTreeNode.FunctionCall(typedIdentifier, arguments);
            }
            case Expression.PrintlnExpression ignored -> new TypedTreeNode.NotImplemented();
            case Expression.StringLiteral stringLiteral ->
                    new TypedTreeNode.StringLiteral(stringLiteral.value(), new FunType.StringType());
        };
    }

    private static TypedTreeNode.ArithmeticExpression anonotateTypesAtArithmeticExpression(
            Expression.ArithmeticExpression arithmeticExpression
    ) {
        return switch (arithmeticExpression) {
            case Expression.BinaryArithmeticExpression binaryArithmeticExpression -> {
                TypedTreeNode.ArithmeticExpression.Operation operation =
                        switch (binaryArithmeticExpression.operation()) {
                            case ADD -> TypedTreeNode.ArithmeticExpression.Operation.ADD;
                            case SUBTRACT -> TypedTreeNode.ArithmeticExpression.Operation.SUBTRACT;
                        };
                yield new TypedTreeNode.BinaryArithmeticExpression(
                        new FunType.TypeVariable(),
                        annotateTypesAtTerm(binaryArithmeticExpression.left()),
                        annotateTypesAtTerm(binaryArithmeticExpression.right()),
                        operation
                );
            }
            case Expression.Operand operand ->
                    new TypedTreeNode.Operand(new FunType.TypeVariable(), annotateTypesAtTerm(operand.left()));
        };
    }

    private static TypedTreeNode.Term annotateTypesAtTerm(Expression.Term term) {
        return switch (term) {
            case Expression.BinaryTerm binaryTerm -> {
                TypedTreeNode.Term.Operation operation = switch (binaryTerm.operation()) {
                    case DIVIDE -> TypedTreeNode.Term.Operation.DIVIDE;
                    case MULTIPLY -> TypedTreeNode.Term.Operation.MULTIPLY;
                };
                yield new TypedTreeNode.BinaryTerm(
                        new FunType.TypeVariable(),
                        annotateTypesAtFactor(binaryTerm.left()),
                        annotateTypesAtFactor(binaryTerm.right()),
                        operation
                );
            }
            case Expression.SimpleTerm simpleTerm ->
                    new TypedTreeNode.SimpleTerm(new FunType.TypeVariable(), annotateTypesAtFactor(simpleTerm.left()));
        };
    }

    private static TypedTreeNode.Factor annotateTypesAtFactor(Expression.Factor factor) {
        return switch (factor) {
            case Expression.IntegerLiteral integerLiteral ->
                    new TypedTreeNode.IntegerLiteral(integerLiteral.literal(), new FunType.IntType());
            case Expression.ParenthesizedExpression parenthesizedExpression -> {
                TypedTreeNode.ArithmeticExpression typedExpression =
                        (TypedTreeNode.ArithmeticExpression) annotateTypesAtExpression(parenthesizedExpression.arithmeticExpression());
                yield new TypedTreeNode.ParenthesizedExpression(typedExpression);
            }
        };
    }

}

File: src/main/scala/com/github/tammo/frontend/type/TypeAnnotate.scala
package com.github.tammo.frontend.`type`

import com.github.tammo.diagnostics.PositionSpan
import com.github.tammo.frontend.`type`.TypedTree.TypedIdentifier
import com.github.tammo.frontend.ast.SyntaxTree
import com.github.tammo.frontend.ast.SyntaxTree.{ArithmeticExpression, Term}

object TypeAnnotate {

  def annotateTypes(
      syntaxTree: SyntaxTree.CompilationUnit
  ): TypedTree.CompilationUnit = {
    val namespaceDeclaration = syntaxTree.namespace
      .map(namespace =>
        TypedTree.NamespaceDeclaration(namespace.identifier, namespace.span)
      )

    TypedTree.CompilationUnit(
      namespaceDeclaration,
      annotateTypeAtClassDeclaration(syntaxTree.classDeclaration),
      syntaxTree.span
    )
  }

  private def annotateTypeAtClassDeclaration(
      classDeclaration: SyntaxTree.ClassDeclaration
  ): TypedTree.ClassDeclaration =
    TypedTree.ClassDeclaration(
      classDeclaration.name,
      classDeclaration.effects.map(annotateTypesAtEffectDeclaration),
      classDeclaration.functions.map(annotateTypesAtFunctionDeclaration),
      classDeclaration.span
    )

  private def annotateTypesAtEffectDeclaration(
      effectDeclaration: SyntaxTree.EffectDeclaration
  ): TypedTree.EffectDeclaration = {
    val typedIdentifier = TypedTree.TypedIdentifier(
      effectDeclaration.name,
      createFunctionType(
        effectDeclaration.returnType,
        effectDeclaration.parameters
      ),
      PositionSpan(
        effectDeclaration.span.fileId,
        effectDeclaration.span.startOffset,
        effectDeclaration.span.startOffset + effectDeclaration.name.length
      )
    )

    TypedTree.EffectDeclaration(
      typedIdentifier,
      effectDeclaration.parameters.map(annotateTypeAtParameter),
      annotateTypeAtExpression(effectDeclaration.body),
      effectDeclaration.span
    )
  }

  private def annotateTypesAtFunctionDeclaration(
      functionDeclaration: SyntaxTree.FunctionDeclaration
  ): TypedTree.FunctionDeclaration = {
    val typedIdentifier = TypedTree.TypedIdentifier(
      functionDeclaration.name,
      createFunctionType(
        functionDeclaration.returnType,
        functionDeclaration.parameters
      ),
      PositionSpan(
        functionDeclaration.span.fileId,
        functionDeclaration.span.startOffset,
        functionDeclaration.span.startOffset + functionDeclaration.name.length
      )
    )

    TypedTree.FunctionDeclaration(
      typedIdentifier,
      functionDeclaration.parameters.map(annotateTypeAtParameter),
      annotateTypeAtExpression(functionDeclaration.body),
      functionDeclaration.span
    )
  }

  private def annotateTypeAtParameter(
      parameter: SyntaxTree.Parameter
  ): TypedTree.Parameter = TypedTree.Parameter(
    TypedTree.TypedIdentifier(
      parameter.identifier,
      Type.fromString(parameter.`type`),
      PositionSpan(
        parameter.span.fileId,
        parameter.span.startOffset,
        parameter.span.startOffset + parameter.identifier.length
      )
    ),
    parameter.span
  )

  private def annotateTypeAtExpression(
      expression: SyntaxTree.Expression
  ): TypedTree.Expression = expression match
    case expression: SyntaxTree.ArithmeticExpression =>
      annotateTypeAtArithmeticExpression(expression)
    case term: SyntaxTree.Term     => annotateTypeAtTerm(term)
    case factor: SyntaxTree.Factor => annotateTypeAtFactor(factor)
    case application: SyntaxTree.FunctionApplication =>
      annotateTypeAtFunctionApplication(application)
    case SyntaxTree.StringLiteral(value, span) =>
      TypedTree.StringLiteral(value, span)
    case SyntaxTree.PrintExpression(expression, span) =>
      TypedTree.PrintExpression(annotateTypeAtExpression(expression), span)
    case SyntaxTree.ParenthesizedExpression(expression, span) =>
      TypedTree.ParenthesizedExpression(
        annotateTypeAtExpression(expression),
        span
      )

  private def annotateTypeAtArithmeticExpression(
      arithmeticExpression: SyntaxTree.ArithmeticExpression
  ): TypedTree.ArithmeticExpression = arithmeticExpression match
    case SyntaxTree.Operand(left, span) =>
      TypedTree.Operand(Type.Variable(), annotateTypeAtTerm(left), span)
    case SyntaxTree.BinaryArithmeticExpression(left, right, operator, span) =>
      val mappedOperator = operator match
        case SyntaxTree.ArithmeticExpression.Operator.ADD =>
          TypedTree.ArithmeticExpression.Operator.ADD
        case SyntaxTree.ArithmeticExpression.Operator.SUBTRACT =>
          TypedTree.ArithmeticExpression.Operator.SUBTRACT
      TypedTree.BinaryArithmeticExpression(
        Type.Variable(),
        annotateTypeAtTerm(left),
        annotateTypeAtExpression(right),
        mappedOperator,
        span
      )

  private def annotateTypeAtTerm(term: SyntaxTree.Term): TypedTree.Term =
    term match
      case SyntaxTree.UnaryTerm(left, span) =>
        TypedTree.UnaryTerm(Type.Variable(), annotateTypeAtFactor(left), span)
      case SyntaxTree.BinaryTerm(left, right, operator, span) =>
        val mappedOperator = operator match
          case SyntaxTree.Term.Operator.MULTIPLY =>
            TypedTree.Term.Operator.MULTIPLY
          case SyntaxTree.Term.Operator.DIVIDE => TypedTree.Term.Operator.DIVIDE

        TypedTree.BinaryTerm(
          Type.Variable(),
          annotateTypeAtFactor(left),
          annotateTypeAtExpression(right),
          mappedOperator,
          span
        )

  private def annotateTypeAtFactor(
      factor: SyntaxTree.Factor
  ): TypedTree.Factor = factor match
    case SyntaxTree.IntLiteral(literal, span) =>
      TypedTree.IntLiteral(literal, span)

  private def annotateTypeAtFunctionApplication(
      functionApplication: SyntaxTree.FunctionApplication
  ): TypedTree.FunctionApplication =
    TypedTree.FunctionApplication(
      TypedIdentifier(
        functionApplication.identifier,
        Type.Variable(),
        PositionSpan(
          functionApplication.span.fileId,
          functionApplication.span.startOffset,
          functionApplication.span.startOffset + functionApplication.identifier.length
        )
      ),
      functionApplication.arguments.map(annotateTypeAtExpression),
      functionApplication.span
    )

  private def createFunctionType(
      returnType: Option[String],
      parameters: Seq[SyntaxTree.Parameter]
  ): Type = {
    val mappedReturnType =
      returnType.map(Type.fromString).getOrElse(Type.Variable())
    parameters.foldRight(mappedReturnType) { (arg, acc) =>
      Type.FunctionType(Type.fromString(arg.`type`), acc)
    }
  }

}
